tools/cli/vz/cmd/login/login.go

```
@@ -0,0 +1,336 @@
+		verrazzanoAPIURL,
+		caData,
+	)
+
+	// Add the logged-in user with nickname verrazzano
+	helpers.SetUserInKubeConfig("verrazzano",
+		jwtData["access_token"].(string),
+		jwtData["refresh_token"].(string),
+		int64(jwtData["expires_in"].(float64))+time.Now().Unix()-10,
+		int64(jwtData["refresh_expires_in"].(float64))+time.Now().Unix()-10,
+	)
+
+	// Add new context with name verrazzano@oldcontext
+	// This context uses verrazzano cluster and logged-in user
+	// We need oldcontext to fall back after logout
+	helpers.SetContextInKubeConfig("verrazzano"+"@"+helpers.GetCurrentContextFromKubeConfig(),
+		"verrazzano",
+		"verrazzano",
+	)
+
+	// Switch over to new context
+	helpers.SetCurrentContextInKubeConfig("verrazzano" + "@" + helpers.GetCurrentContextFromKubeConfig())
+
+	fmt.Fprintln(streams.Out, "Login successful!")
+	return nil
+}
+
+var authCode = "" //	Http handle fills this after keycloak authentication
+
+// A function to put together all the requirements of authorization grant flow
+// Returns the final jwt token as a map
+func authFlowLogin(caData []byte) (map[string]interface{}, error) {
+
+	// Obtain a available port in non-well known port range
+	listener := getFreePort()
+
+	// Generate random code verifier and code challenge pair
+	codeVerifier, codeChallenge := helpers.GenerateRandomCodePair()
+
+	// Generate the redirect uri using the port obtained
+	redirectURI := helpers.GenerateRedirectURI(listener)
+
+	// Generate the login keycloak url by passing the required url parameters
+	loginURL := helpers.GenerateKeycloakAPIURL(codeChallenge,
+		redirectURI,
+	)
+
+	// Busy wait when the authorization code is still not filled by http handle
+	// Close the listener once we obtain it
+	go func() {
+		for authCode == "" {
+
+		}
+		listener.Close()
+	}()
+
+	// Make sure the go routine is running
+	time.Sleep(time.Second)
+
+	// Open the generated keycloak login url in the browser
+	err := helpers.OpenURLInBrowser(loginURL)
+	if err != nil {
```

return error rather than continuing

---

okay